src/webp/decoder.rs

```diff
@@ -37,44 +72,38 @@ impl<R: Read> WebPDecoder<R> {
     }
 
     fn read_riff_header(&mut self) -> ImageResult<u32> {
-        let mut riff = Vec::with_capacity(4);
-        self.r.by_ref().take(4).read_to_end(&mut riff)?;
-        let size = self.r.read_u32::<LittleEndian>()?;
-        let mut webp = Vec::with_capacity(4);
-        self.r.by_ref().take(4).read_to_end(&mut webp)?;
-
-        if &*riff != b"RIFF" {
-            return Err(ImageError::Decoding(DecodingError::with_message(
-                ImageFormat::WebP.into(),
-                "Invalid RIFF signature",
-            )));
+        let mut riff = [0; 4];
+        self.r.read_exact(&mut riff)?;
+        if &riff != b"RIFF" {
+            return Err(DecoderError::RiffSignatureInvalid(riff).into());
         }
 
-        if &*webp != b"WEBP" {
-            return Err(ImageError::Decoding(DecodingError::with_message(
-                ImageFormat::WebP.into(),
-                "Invalid WEBP signature",
-            )));
+        let size = self.r.read_u32::<LittleEndian>()?;
+
+        let mut webp = [0; 4];
+        self.r.read_exact(&mut webp)?;
+        if &webp != b"WEBP" {
+            return Err(DecoderError::WebpSignatureInvalid(webp).into());
         }
 
         Ok(size)
     }
 
     fn read_vp8_header(&mut self) -> ImageResult<u32> {
         loop {
-            let mut chunk = Vec::with_capacity(4);
-            self.r.by_ref().take(4).read_to_end(&mut chunk)?;
+            let mut chunk = [0; 4];
+            self.r.read_exact(&mut chunk)?;
```

Comment:
This is a bugfix as well, right? Because before it could silently truncate the chunk name by a temporarily exhausted stream. Not sure how one could test for that bug though.

Reply:
I *think* the `take()`/`read_to_end()` combo worked, too:
> All bytes read from this source will be appended to the specified buffer. This function will continuously call `read()` to append more data until `read()` returns either `Ok(0)` or an error of non-`ErrorKind::Interrupted` kind.

Which sounds equivalent to `read_exact()` when paired with `take()`, but I'm knackered and might very well be missing something.

Reply:
It may read _less_ bytes than `take` if the input is shorter. Since `Ok(0)` might be a temporary condition, it's only a convention to interpret it as eof, it _could_ in theory stop reading and still succeed in the following calls. In any case, it would silently succeed and return an erro from an incorrect line.

Reply:
Oh, indeed! Will add that to the commit message.